### PR TITLE
feat(agent): Allow separators for namepass and namedrop filters

### LINKED ISSUE #14056: Allow specifying separators while using glob match for filtering
### Use Case

I'm using [[inputs.socket_listener]] plugin to collect graphite-style metrics. The problem comes while using filters like namepass/namedrop where wildcard includes the graphite measurement separator `.`. So there is no way to explicitly say, pass all metrics where 2nd measurement is anything but 3rd measurement is 'B'. The only way of saying this is `A.*.B` but this implicitly allows metrics like `A.X.Y.B`. Browsing through the code, seems like this type of match is supported by the underlying glob match library.

Requesting here to add a way to explicitly specify separator inside config keys like namepass, namedrop, tagpass, tagdrop; so that glob match can use those separators to be not included in the wildcard match.

### Expected behavior

given the following input plugin:

```
[[inputs.socket_listener]]
  ## URL to listen on
  service_address = "<>"
  data_format = "graphite"

  templates = [
    "measurement*",
  ]

  namepass = [
    "A.*.B"
  ]
```

`A.C.B` should pass
`A.C.D.B` should *not* pass

### Actual behavior

Both `A.C.B` & `A.C.D.B` pass.

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by akash-rubrik**
# Required for all PRs

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #14056

### CODE REVIEW

**Comment by akash-rubrik** [comment]
!signed-cla

**Comment by akash-rubrik** [comment]
@srebhan can you help review this?

File: filter/filter.go [hunk 1]
```diff
@@ -12,24 +12,35 @@ type Filter interface {
...
 	}
 
 	// check if we can compile a non-glob filter
-	noGlob := true
-	for _, filter := range filters {
-		if hasMeta(filter) {
-			noGlob = false
-			break
+	noGlob := len(separators) == 0
+	if noGlob {
```
Comment (srebhan) [inline filter/filter.go:38]: The `if` is not necessary.

File: config/config.go [hunk 2]
```diff
@@ -1582,6 +1584,19 @@ func (c *Config) getFieldInt64(tbl *ast.Table, fieldName string, target *int64)
...
+			str, ok := kv.Value.(*ast.String)
+			if !ok {
+				c.addError(tbl, fmt.Errorf("found unexpected format while parsing %q, expecting string format", fieldName))
+				return
+			}
+			*target = []rune(str.Value)
+		}
+	}
+}
+
```
Comment (srebhan) [inline config/config.go:1599]: How about handling this as "strings" in the settings and convert to runes in the filter? This way we can directly infer from this code what to expect in the settings...

**Review by srebhan — COMMENTED** [review]
Thanks for the nice and comprehensive PR @akash-rubrik! Especially for the unit-test!!! Just have two minor comments...

File: config/config.go (hunk 2 above)
Comment (akash-rubrik) [reply config/config.go:1599]: @srebhan Do you mean to remove the above method and just use the below snippets instead?

```GO
c.getFieldString(tbl, "namepass_separator", &f.NamePassSeparators)

...

type Filter struct {
    NameDropSeparators string
...

func (f *Filter) Compile() error {
    .
    f.nameDropFilter, err = filter.Compile(f.NameDrop, []rune(f.NameDropSeparators)...)
    .
}
```

File: filter/filter.go (hunk 1 above)
Comment (akash-rubrik) [reply filter/filter.go]: done

File: config/config.go (hunk 2 above)
Comment (akash-rubrik) [reply config/config.go]: have made the above changes, let me know if I misunderstood.

File: config/testdata/single_plugin.toml [hunk 3]
```diff
@@ -1,7 +1,9 @@
 [[inputs.memcached]]
   servers = ["localhost"]
   namepass = ["metricname1"]
+  namepass_separator = "."
   namedrop = ["metricname2"]
+  namedrop_separator = "."
   fieldpass = ["some", "strings"]
```
Comment (srebhan) [inline config/testdata/single_plugin.toml:7]: Can you please leave existing tests as they were!? We usually use them as a guarantee for backward-compatibility... Same for the other toml files below. If you want to test something, just add a new config or test.

File: docs/CONFIGURATION.md [hunk 4]
```diff
@@ -626,11 +626,15 @@ sent onwards to the next stage of processing.
 
 - **namepass**:
 An array of [glob pattern][] strings. Only metrics whose measurement name
-matches a pattern in this list are emitted.
+matches a pattern in this list are emitted. Additionally, custom list of
+separators can be specified using **namepass_separator**. These separators
```
Comment (srebhan) [inline docs/CONFIGURATION.md:630]:
```suggestion
separators can be specified using `namepass_separator`. These separators
```

File: docs/CONFIGURATION.md [hunk 5]
```diff
@@ -626,11 +626,15 @@ sent onwards to the next stage of processing.
...
+are excluded from wildcard glob pattern matching.
 
 - **namedrop**:
 The inverse of `namepass`. If a match is found the metric is discarded. This
-is tested on metrics after they have passed the `namepass` test.
+is tested on metrics after they have passed the `namepass` test. Additionally,
+custom list of separators can be specified using **namedrop_separator**. These
```
Comment (srebhan) [inline docs/CONFIGURATION.md:636]:
```suggestion
custom list of separators can be specified using `namedrop_separator`. These
```

**Review by srebhan — COMMENTED** [review]
Thanks for the update @akash-rubik! This was exactly what I meant. I have two small formatting comments and a request to leave existing tests as-is to also test backward-compatibility. We are very close IMO...

File: config/testdata/single_plugin.toml (hunk 3 above)
Comment (akash-rubrik) [reply config/testdata/single_plugin.toml]: makes sense, done

**Review by Hipska — APPROVED** [review]
Thanks for this feature. The only thing I'm wondering is if there is any scenario you would have a different separator for Pass or for Drop? It looks a bit too much to have 2 separate config items for that..

**Review by srebhan — APPROVED** [review]
Awesome! Thanks for your contribution @akash-rubrik!

**Review by powersj — APPROVED** [review]
Thank you!

**Pull request merged**